[AIE] Internalize non-kernel symbols.

SagarMaheshwari99 opened the pull request:
This saves compilation time, size of the final
executable, and size of any intermediate dumps.

Run Internalize pass early in the opt pipeline followed by global DCE pass. To enable it RT can pass -aie-internalize-symbols option.

Edit: Also some refactoring.

andcarminati commented:
Hi @SagarMaheshwari99, is there any other specific use apart of time/space reduction? I think that this cause problems at linking stage.

SagarMaheshwari99 commented:
> Hi @SagarMaheshwari99, is there any other specific use apart of time/space reduction? I think that this cause problems at linking stage.

@andcarminati The current use is only time/space reduction. Can you please elaborate on what kind of problems can be caused at the linking stage?

martien-de-jong commented on llvm/lib/Target/AIE/CMakeLists.txt:
```diff
@@ -114,6 +114,7 @@ add_llvm_target(AIECodeGen
    Core
    CodeGen
    CodeGenTypes
+   IPO
```
Check: this is the library where the internalize pass lives?

SagarMaheshwari99 replied:
Yes, and GlobalDCE.

martien-de-jong commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -82,6 +90,15 @@ AIE2TargetMachine::AIE2TargetMachine(const Target &T, const Triple &TT,
+/// Predicate for Internalize pass.
+static bool mustPreserveGV(const GlobalValue &GV) {
+  if (const Function *F = dyn_cast<Function>(&GV))
+    return F->isDeclaration() || F->getName().starts_with("_main");
```
I just pushed a PR to change this to "main" in the startup code. I don't know the rationale for _main, nor did @konstantinschwarz. Everything would look more familiar if we could just use main, like the rest of the world.

andcarminati replied:
Hi, maybe we can have a option like this for flexibility:
```
cl::opt<std::string> EntryFunctionName ...
```

konstantinschwarz replied:
Having an option for this sounds good! It could default to `main`, but would allow users to specify their own entry function

konstantinschwarz replied:
Instead of `.starts_with`, change to `equals`? Otherwise we might get false positives

konstantinschwarz replied:
Instead of .starts_with, change to equals? Otherwise we might get false positives

SagarMaheshwari99 replied:
`.starts_with` was also useful for `_main_init`... so I didn't change it. But I think a CLI option to add a list of functions is convenient... 

@martien-de-jong : The _main was used because "our internal linker" expected it (somewhere while creating object, the internal compiler converted main to _main) and we had no way other than use _main directly in the cc file. I agree to change it conditionally for the lld flow back to `main` in aiecompiler. 

I am thinking of also removing the extern "C", since we should be able to handle the mangling? Any opinions?

SagarMaheshwari99 replied:
Made the changes.

konstantinschwarz replied:
I think its sensible for the `FunctionSkipList` to contain `main` by default. Now we require both options to be set

andcarminati commented:
> > Hi @SagarMaheshwari99, is there any other specific use apart of time/space reduction? I think that this cause problems at linking stage.
> 
> @andcarminati The current use is only time/space reduction. Can you please elaborate on what kind of problems can be caused at the linking stage?

Hi @SagarMaheshwari99 , I was thinking in a work flow where the final executable Elf would be created by linking several different object files with different symbols. If this option is off by default then the user can use it for self contained CUs. For space reduction, another alternative we could think is the default use of `-ffunction-sections` and `-fdata-sections` in the frontend plus linker garbage collection.

SagarMaheshwari99 commented:
> > > Hi @SagarMaheshwari99, is there any other specific use apart of time/space reduction? I think that this cause problems at linking stage.
> > 
> > 
> > @andcarminati The current use is only time/space reduction. Can you please elaborate on what kind of problems can be caused at the linking stage?
> 
> Hi @SagarMaheshwari99 , I was thinking in a work flow where the final executable Elf would be created by linking several different object files with different symbols. If this option is off by default then the user can use it for self contained CUs. For space reduction, another alternative we could think is the default use of `-ffunction-sections` and `-fdata-sections` in the frontend plus linker garbage collection.

Hi @andcarminati, Yes, we should enable function and data sections in frontend and garbage collection in the linker by default. But the internalize pass was also useful to save up on the compile time plus help make debugging less noisy...

andcarminati commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -64,6 +66,17 @@ static cl::opt<bool> EnablePreMISchedCoalescer(
+    cl::init(false), cl::Hidden);
+// Option to skip the functions we don't want to internalize.
+static cl::list<std::string>
+    FunctionSkipList("aie2-internalize-skip-function",
```
Maybe `aie2-internalize-skip-functions`

andcarminati commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -64,6 +66,17 @@ static cl::opt<bool> EnablePreMISchedCoalescer(
+// Option to skip the functions we don't want to internalize.
+static cl::list<std::string>
+    FunctionSkipList("aie2-internalize-skip-function",
+                     cl::desc("List of symbols to skip internalization"),
```
I think it is better "List of function names to skip internalization".

andcarminati commented on llvm/test/CodeGen/AIE/aie2/internalize.ll:
```diff
@@ -0,0 +1,58 @@
+; RUN: opt -O1 -S -mtriple=aie2-none-unknown-elf -aie2-internalize-symbols -aie2-internalize-skip-function="_main" < %s | FileCheck %s
```
I think we can have an additional test considering two function in the list.

SagarMaheshwari99 replied:
Added..

konstantinschwarz commented on llvm/test/CodeGen/AIE/aie2/internalize.ll:
```diff
@@ -0,0 +1,123 @@
+; RUN: opt -O1 -S -mtriple=aie2-none-unknown-elf -aie2-internalize-symbols < %s | FileCheck --check-prefix=ONLYMAIN %s
+; RUN: opt -O1 -S -mtriple=aie2-none-unknown-elf -aie2-internalize-symbols -aie2-internalize-skip-functions="skipfunc1" -aie2-internalize-skip-functions="skipfunc2" -aie2-internalize-skip-functions="skipfunc3" < %s | FileCheck --check-prefix=FUNCLIST %s
```
Maybe use `cl::CommaSeparated` to keep the command line shorter?
i.e. allow to pass the symbols as
```
-aie2-internalize-skip-functions="skipfunc1,skipfunc2,skipfunc3"
```

gbossu commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -64,6 +66,17 @@ static cl::opt<bool> EnablePreMISchedCoalescer(
+static cl::opt<bool> InternalizeSymbols(
+    "aie2-internalize-symbols",
+    cl::desc("Enable elimination of non-kernel functions and unused globals"),
+    cl::init(false), cl::Hidden);
```
Do we have a common pass pipeline that can be shared between different AIE versions?

SagarMaheshwari99 replied:
I am not sure. They ultimately derive from TargetPassConfig, so maybe we could have a generic one...

gbossu replied:
Maybe have a chat with @abnikant ? It would be sad to just duplicate all those files for every AIE version.

abnikant replied:
In the aie-private branch the AIE2PPassConfig derives from AIE2PassConfig. Most of passes are shared between AIE2 and AIE2p.
class AIE2PassConfig : public AIEPassConfig {
}
class AIEPassConfig : public TargetPassConfig {
}

I suppose you can setup your pass in AIEPassConfig.

SagarMaheshwari99 replied:
Refactored the code to be in AIETargetMachine.cpp

gbossu commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -82,6 +96,19 @@ AIE2TargetMachine::AIE2TargetMachine(const Target &T, const Triple &TT,
   setGlobalISelAbort(GlobalISelAbortMode::Enable);
 }
 
+/// Predicate for Internalize pass.
```
Nit: I feel the comment does not help much. Instead, maybe give a rationale about what should be preserved and why?

gbossu commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -64,6 +66,18 @@ static cl::opt<bool> EnablePreMISchedCoalescer(
+static cl::list<std::string>
+    FunctionSkipList("aie2-internalize-skip-functions",
+                     cl::desc("List of function names to skip internalization"),
+                     cl::Hidden, cl::list_init<std::string>({"main"}),
```
We sometimes also have a `_main` symbol. I guess we don't have to include it here because `main` is the real entry point?

SagarMaheshwari99 replied:
When using the peano flow, the `main` will be the "real" entry point(from today). In our other flow, we have `_main` as the entry point bcoz that's what the linker expects... For internalizing _main, we need to pass it on the command line. This will override `main` as the entry point.

gbossu commented on llvm/test/CodeGen/AIE/aie2/internalize.ll:
```diff
@@ -0,0 +1,123 @@
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @skipfunc1()
+; ONLYMAIN-NOT: tail call fastcc void @func_used_skipfunc1
```
Maybe just `call{{.*}}void @func_used_skipfunc1`? I'm nervous about all those attributes that can change when we bump the LLVM version. (Same for other check lines)

SagarMaheshwari99 replied:
yeah i should probably change that.

gbossu commented on llvm/test/CodeGen/AIE/aie2/internalize.ll:
```diff
@@ -0,0 +1,123 @@
+    %val = add i32 42, 0
+    tail call fastcc void @func_used_skipfunc3(ptr addrspace(1) @gvar_used, i32 %val)
+    ret void
+}
```
Maybe add a test where the address of a function escapes (e.g. it gets stored somewhere)?

gbossu commented on llvm/lib/Target/AIE/AIETargetMachine.cpp:
```diff
@@ -43,13 +43,27 @@
+    "aie-internalize-symbols",
+    cl::desc("Enable elimination of non-kernel functions and unused globals"),
+    cl::init(false), cl::Hidden);
+// Option to skip the functions we don't want to internalize.
```
Nit: Skip a line for readability?

SagarMaheshwari99 replied:
done

gbossu commented on llvm/lib/Target/AIE/AIETargetMachine.cpp:
```diff
@@ -261,6 +275,26 @@ void AIEBaseTargetMachine::registerDefaultAliasAnalyses(AAManager &AAM) {
+/// (default entry point is 'main'). Preserve function declarations, as if we
+/// internalize a declaration, it would not be accessible outside of its
+/// translation unit, which could lead to linker errors if the function is
+/// used/defined elsewhere.
```
I think the comment is a bit convoluted, maybe simply write `Preserve functions that can be an entry point or that have uses within the Module`?

SagarMaheshwari99 replied:
done

gbossu commented on llvm/test/CodeGen/AIE/internalize.ll:
```diff
@@ -0,0 +1,126 @@
+; FUNCLIST-NOT: call{{.*}}void @func_used_noinline
+; FUNCLIST-NOT: store volatile
+; FUNCLIST-NOT: call{{.*}}void @decl_func()
+; FUNCLIST-NOT: ret void
```
Nit: I think just `; FUNCLIST-NOT: define{{.*}}void @main()` is fine

SagarMaheshwari99 replied:
done

gbossu commented on llvm/test/CodeGen/AIE/internalize.ll:
```diff
@@ -0,0 +1,126 @@
+  ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @skipfunc3()
```
Nit: Same, I think when we expect a function **not** to be there, something like `CHECK-NOT: define{{.*}} @thefunction` is enough

SagarMaheshwari99 replied:
Done

gbossu reviewed (dismissed):
LGTM, please address comments as you please. Did you verify QoR? I'd be curious about PM size reduction.

SagarMaheshwari99 commented:
> LGTM, please address comments as you please. Did you verify QoR? I'd be curious about PM size reduction.

I did run it a while back but I lost the results, I will run it again.

SagarMaheshwari99 committed:
[AIE] Internalize non-kernel symbols.

This saves compilation time, size of the final
executable, and size of any intermediate dumps.

Run Internalize pass early in the opt pipeline followed by global
DCE pass. To enable it RT can pass -aie-internalize-symbols option.

gbossu commented on llvm/lib/Target/AIE/AIETargetMachine.cpp:
```diff
@@ -38,11 +39,32 @@
+static cl::opt<bool> InternalizeSymbols(
+    "aie-internalize-symbols",
+    cl::desc("Enable elimination of non-kernel functions and unused globals"),
+    cl::init(false), cl::Hidden);
```
As discussed offline, I think this should be enabled by default.

gbossu replied:
Discussed offline, static libraries can be a problem. We should only internalize and simplify if we are producing an "final executable". Let's keep it disabled by default.

SagarMaheshwari99 committed:
[AIE] Bring EnableCustomAliasAnalysis to AIETargetMachine.

konstantinschwarz reviewed (approved):
All downstream issues got resolved, let's land this!
